### PR TITLE
[#715] Support the RocketMQ TableSource based on the legacy Source implementation

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

At present, RocketMQ connector supports the implementation of legacy SourceFunction interface, which needs the `ScanTableSource` implementation to expand Table/SQL capabilities. RocketMQ connector should support the RocketMQ TableSource based on the legacy SourceFunction interface.

## Brief changelog

 - Support `RocketMQScanTableSource` based on the implementation `RocketMQSourceFunction` of legacy `SourceFunction` interface.

## Verifying this change

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/legacy/common/serialization/RowKeyValueDeserializationSchema.java [hunk 1]
```diff
@@ -0,0 +1,415 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * <p>http://www.apache.org/licenses/LICENSE-2.0
```
Comment (vongosling) [inline rocketmq-flink/src/main/java/org/apache/rocketmq/flink/legacy/common/serialization/RowKeyValueDeserializationSchema.java:8]: Do not format the license style, you could make a change in ide,https://www.jetbrains.com/help/idea/copyright.html#update-copyright

**Comment by vongosling** [comment]
@duhenglucky @RongtongJin @SteNicholas This connector may donate to flink like this, https://github.com/apache/flink/pull/15304. I would like to help if we could step it.

**Comment by duhenglucky** [comment]
> @duhenglucky @RongtongJin @SteNicholas This connector may donate to flink like this, [apache/flink#15304](https://github.com/apache/flink/pull/15304). I would like to help if we could step it.

this is also one of @SteNicholas's targets in the near future.

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/legacy/common/serialization/RowKeyValueDeserializationSchema.java [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+                                    + "]");
+                    lastLogHandleFieldTime = now;
+                }
+                return null;
+            case SKIP_SILENT:
+                return null;
+            case CUT:
+            case NULL:
+            case PAD:
+                {
```
Comment (duhenglucky) [inline rocketmq-flink/src/main/java/org/apache/rocketmq/flink/legacy/common/serialization/RowKeyValueDeserializationSchema.java:257]: Keep the code format consistent with other branches

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/legacy/common/serialization/RowKeyValueDeserializationSchema.java [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+                    return data;
+                }
+            case EXCEPTION:
+                throw new RuntimeException();
```
Comment (duhenglucky) [inline rocketmq-flink/src/main/java/org/apache/rocketmq/flink/legacy/common/serialization/RowKeyValueDeserializationSchema.java:261]: There seems to be a need for a logger to print what happened, and this exception does not contain any information

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/legacy/common/serialization/RowKeyValueDeserializationSchema.java [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+                return null;
+            case SKIP_SILENT:
+                return null;
+            default:
+            case CUT:
+            case NULL:
+            case PAD:
+                return data;
+            case EXCEPTION:
+                throw new RuntimeException();
```
Comment (duhenglucky) [inline rocketmq-flink/src/main/java/org/apache/rocketmq/flink/legacy/common/serialization/RowKeyValueDeserializationSchema.java:291]: same as upper comment

File: rocketmq-flink/pom.xml [hunk 5]
```diff
@@ -34,7 +34,7 @@
         <maven.compiler.source>1.8</maven.compiler.source>
         <maven.compiler.target>1.8</maven.compiler.target>
         <rocketmq.version>4.7.1</rocketmq.version>
-        <flink.version>1.13.0</flink.version>
+        <flink.version>1.13.1</flink.version>
```
Comment (ni-ze) [inline rocketmq-flink/pom.xml:37]: Why rocketmq-flink package has flink jar?That will lead a package conflict when run on Flink.

image

It should set scope to "provided".

```xml
		<dependency>
			<groupId>org.apache.flink</groupId>
			<artifactId>flink-streaming-java_${scala.binary.version}</artifactId>
			<version>${project.version}</version>
			<scope>provided</scope>
		</dependency>
		<dependency>
			<groupId>org.apache.flink</groupId>
			<artifactId>flink-java</artifactId>
			<version>${project.version}</version>
			<scope>provided</scope>
		</dependency>
```

File: rocketmq-flink/src/test/java/org/apache/rocketmq/flink/source/table/RocketMQDynamicTableSourceFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            new ResolvedSchema(
+                    Collections.singletonList(Column.physical("name", STRING().notNull())),
+                    new ArrayList<>(),
+                    null);
+
+    private static final String IDENTIFIER = "rocketmq";
+    private static final String TOPIC = "test_source";
+    private static final String CONSUMER_GROUP = "test_consumer";
+    private static final String NAME_SERVER_ADDRESS =
+            "http://${instanceId}.${region}.mq-internal.aliyuncs.com:8080";
```
Comment (duhenglucky) [inline rocketmq-flink/src/test/java/org/apache/rocketmq/flink/source/table/RocketMQDynamicTableSourceFactoryTest.java:58]: Please do not include any vendor-specific code

**Comment by SteNicholas** [comment]
> @duhenglucky @RongtongJin @SteNicholas This connector may donate to flink like this, [apache/flink#15304](https://github.com/apache/flink/pull/15304). I would like to help if we could step it.

@vongosling IMO, RocketMQ connector couldn't contribute to Flink connectors in a short time. It is recommended to graduate this connector first.

**Comment by ni-ze** [comment]
@SteNicholas 
https://github.com/apache/rocketmq-externals/issues/786#issue-973254402

**Comment by ni-ze** [comment]
@SteNicholas 
MessageQueues never update duration Flink job running, How to support message queues rebalance?

https://github.com/apache/rocketmq-externals/issues/787#issue-973258768

**Comment by SteNicholas** [comment]
@ni-ze , thanks for the reporter to the problems of `RocketMQSource`. The problems of `RocketMQSource` could be fixed in another pull request.

File: rocketmq-flink/pom.xml [hunk 7]
```diff
@@ -45,11 +45,13 @@
             <groupId>org.apache.flink</groupId>
             <artifactId>flink-java</artifactId>
             <version>${flink.version}</version>
+            <scope>provided</scope>
```
Comment (vongosling) [inline rocketmq-flink/pom.xml:48]: Good polish

**Review by vongosling — APPROVED** [review]

**Comment by SteNicholas** [comment]
> @SteNicholas
> MessageQueues never update duration Flink job running, How to support message queues rebalance?
> 
> [#787 (comment)](https://github.com/apache/rocketmq-externals/issues/787#issue-973258768)

@ni-ze , in legacy `RocketMQSource`, the `MessageQueue` doesn't support to rebalance during the Flink job running.

**Comment by vongosling** [comment]
@SteNicholas I see the status of our flink integration, there is still improvement to polish before we call for gruduation. #787 (comment), do we have a plan to support this, like the partition discovery feature in here. https://ci.apache.org/projects/flink/flink-docs-release-1.4/dev/connectors/kafka.html#kafka-consumers-topic-and-partition-discovery

**Review by duhenglucky — APPROVED** [review]

**Pull request merged**